keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -1109,21 +1325,16 @@ Below are some examples to consider, in addition to the aforementioned [maturity
 
 - Address reviews and bug reports from Alpha users
 - E2e tests are in Testgrid and linked in KEP
-- A scalability test to demonstrate the limited impact of the additional API call
-  when terminating a Pod
-- Re-evaluate modification to kubelet to send a dedicated condition when
-  terminating a Pod, based on user feedback (see: [Termination initiated by Kubelet](#termination-initiated-by-kubelet))
-- Re-evaluate supporting of `onExitCodes` when `restartPolicy=OnFailure` (see: [Relationship with Pod.spec.restartPolicy](#relationship-with-podspecrestartpolicy))
-- Re-evaluate introduction of a generic opinionated condition type
-  indicating that a pod should be retried (see: [Evolving condition types](#evolving-condition-types))
+- implementation of adding a dedicated pod condition by Kubelet when
+  terminating a Pod (see: [Termination initiated by Kubelet](#termination-initiated-by-kubelet))
 - Simplify the code in job controller responsible for detection of failed pods
   based on the fix for pods stuck in the running phase (see: [Marking pods as Failed](marking-pods-as-failed)).
   Also, introduce a worker in the disruption controller to mark pods
   stuck in the pending phase, with set deletionTimestamp, as failed.
 - Commonize the code for appending pod conditions between components
 - Do not update the pod disruption condition (with type=`DisruptionTarget`) if
   it is already present with `status=True`
-- Review and implement if feasible adding of pod conditions with the use of
+- Re-evaluate and implement if feasible adding of pod conditions with the use of
```

Comment:
@lavalamp @liggitt @alculquicondor  question regarding the use of SSA for appending the disruption conditions for pod failures which is added as a graduation Beta based on discussions: https://github.com/kubernetes/kubernetes/pull/110959/commits/04fcbd721cd3e72eb9375275891e9bc3e28c3341#r929159947 and https://github.com/kubernetes/kubernetes/pull/110959#issuecomment-1199481270. Do you think we keep it as a graduation criteria for Beta or it could be ticketed as an independent PR? I’m asking because there are already a number of things planned for Beta and applying SSA would also substantial changes to unit tests.

Reply:
Are we currently using Patch or Update? I'd want to make sure we are using patch before graduating to beta to avoid needing to handle retries on the client side, but I don't feel strongly about whether SSA patch is used

Reply:
We use Patch:
https://github.com/kubernetes/kubernetes/blob/661899f1a14680840e06c84bc3ff46f73d203012/pkg/controller/nodelifecycle/scheduler/taint_manager.go#L135
https://github.com/kubernetes/kubernetes/blob/661899f1a14680840e06c84bc3ff46f73d203012/pkg/controller/podgc/gc_controller.go#L339
https://github.com/kubernetes/kubernetes/blob/661899f1a14680840e06c84bc3ff46f73d203012/pkg/scheduler/framework/preemption/preemption.go#L356
only from within apiserver we call Update:
https://github.com/kubernetes/kubernetes/blob/661899f1a14680840e06c84bc3ff46f73d203012/pkg/registry/core/pod/storage/eviction.go#L315

Reply:
sgtm

Reply:
Patch doesn't prevent retries depending on how you do it. The patch computation I clicked through to appears to include the RV, which means there will be retries needed. (assuming it started with an object with RV set, which I don't know for sure).

If you don't include the RV, then it's very difficult to produce a patch that will reference a specific condition, since they are a named list.

I think you will have to do SSA to get the property Jordan asked for. But I'm not actually sure that property is super important.

(However if retries are needed you need to do them or it seems like this feature won't work correctly.)

Reply:
Actually, it looks like the resource version is trimmed https://github.com/kubernetes/kubernetes/blob/661899f1a14680840e06c84bc3ff46f73d203012/pkg/util/pod/pod.go#L49

Reply:
You are right.

What prevents the resulting patch from modifying the wrong condition if one is added/removed in a race?

Reply:
I don't think there is anything that prevents it. I tested the following scenario in integration tests:
- add condition with type: "ConditionType1" 
- generate a patch (1) to delete the "ConditionType1" condition
- generate a patch (2) to add a condition with type: "ConditionType2"
- apply the patch (2) to add the condition "ConditionType2"
- apply the patch (1) to delete the "ConditionType1"
Result (issue): no conditions are present, even though we never send a patch to delete "ConditionType2".

So, I think we should solve it. 
The question remains if this is a graduation criteria for Beta or GA @lavalamp @liggitt? I keep it for Beta for now.

Reply:
Thanks for investigating. I think you should fix this for beta. SSA is probably the easiest way, but I won't be picky about how you fix it.